src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalyzer.cs

```diff
@@ -12,6 +15,23 @@ namespace Microsoft.CodeAnalysis.Diagnostics
     /// </summary>
     public abstract class DiagnosticAnalyzer
     {
+        /// <summary>
+        /// This field caches the result of <see cref="AnalysisContext.ConfigureGeneratedCodeAnalysis"/> for the first
+        /// initialization containing at least one registered callback.
+        /// </summary>
+        /// <value>
+        /// The value is a <see cref="GeneratedCodeAnalysisFlags"/> stored as an integer, or -1 if the value has not
+        /// been assigned.
+        /// </value>
+        private int _generatedCodeAnalysisFlags = -1;
+
+        /// <summary>
+        /// This collection is typically <see langword="null"/>. For cases where code analysis flags for a compilation
+        /// differ from <see cref="_generatedCodeAnalysisFlags"/> (only occurs if an analyzer does not consistently
+        /// report this flag), this collection is created to hold the differing values.
+        /// </summary>
+        private ConditionalWeakTable<Compilation, StrongBox<GeneratedCodeAnalysisFlags>>? _generatedCodeAnalysisFlagsOverride;
```

Comment:
⚠️ This pull request is _heavily_ optimized for the case where each call to `DiagnosticAnalyzer.Initialize` produces the same call to `ConfigureGeneratedCodeAnalysis`. I couldn't find any cases where this condition does not hold, but if there are cases in the wild those analyzers could see a non-trivial penalty.

Reply:
do we have tests that can check that this works when that condition doesn't hold?